Gitx dev diff word highlight

muhqu opened the pull request:
Integrated an inline word diff algorithm to highlight changed words in diffs. similar to github's hybrid mode. 
Addressing  #168

The javascript implementation is based on John Resig's javascript diff algorithm described here: http://ejohn.org/projects/javascript-diff-algorithm/

![diff-testfile](https://f.cloud.github.com/assets/26716/967411/7a342be2-0577-11e3-898e-844ea0f24eb2.png)

---

Known Issues
- [x] fix "stage line" behaviour when clicking on highlighted words
- [x] fix invalid diff displayed e.g. just append something at the end of a line
- [x] fix highlight of indention changes e.g. tabs to spaces: need to preserve tabs when passing diff data to javascript, currently tabs are already replaced by spaces
- [x] ~~fix diff view keeps showing "This is a large commit..." for some commits... need to figure out what's happening...~~ invalid

Bonus Todos
- ~~allow diff colors to be customized via `git config`~~ superseded by #249 
- [x] display file renames inline, e.g. very/long/path/to/{ old -> new }-filename.html

muhqu committed:
integrate word changes highlighting in diff.

javascript diff algorithm by John Resig ( http://ejohn.org/ )

muhqu committed:
fix highlight of trailing whitespace

muhqu commented:
There are still some issues with this, which I need to sort out before it's ready to merge... see "Known Issues / Todos" within the description.

atuttle commented:
:+1: That screenshot looks great. How well does it handle indentation changes? It sounds like, after you fix the 2nd to-do item, we'd expect just a green bar showing added indentation, not a pair of red+green lines?

muhqu committed:
fix various inline diff display issues.

- make the inline diff aware of whitespace changes
- added a test page to verify the diff output stays the same
- line selection for stage/unstage in commit dialog

muhqu commented:
as far as I can tell all displaying issues have been fixed now.

@atuttle heres a screenshot showing various changes... unfortunately indention changes (e.g. tabs to spaces) will not yet correctly highlighted because the diff line data I receive in javascript already converted tabs to spaces. so there is no way to detect any changes.
![diff-demo](https://f.cloud.github.com/assets/26716/960241/b9cc54f4-04b0-11e3-8525-037557809c88.png)

atuttle commented:
Actually I'm less concerned with conversion between tabs and spaces and more concerned with addition of more tabs/spaces. For example, the case where I decide to wrap a block with an If statement, or a new div - in this case the content inside the new wrapping statement gets indented but doesn't otherwise change.

It seems, based on lines 16-18 of your screenshotted diff that the support is more or less what I'm looking for. Although I'd rather it be a nice dark green block for added spaces, ala the reddish block for the removed whitespace at the end of line 21. Otherwise, perfect.

atuttle commented:
For what it's worth, I also preferred the **red-on-red-on-red**, and **green-on-green-on-green** of the original screenshot over this new gray-ish background for changes in the new screenshot.

muhqu commented:
Yeah, I changed the colors a bit. First Screenshot already changed the
background colors to be like the ones GitHub is using. But then I felt
the original GitX colors were just fine, so I based the highlighted
background-color on GitX original colors making them a tip darker. I
don't want the highlight colors to be too distracting. The
highlighting should just be a minor hint, while the trailing
whitespace can be seen as a major hint.

Hope this clarifies my intention. But I'm absolutely open to color
suggestions. ;-)

Btw, the GitHub diff uses black text on all colored backgrounds... How
about that?

rowanj commented:
Looks really interesting!

atuttle commented:
I'd be all for making it look more like GitHub... black text with 2 shades of red/green.

muhqu committed:
diff css adjustments

- increase saturation of highlight backgrounds
- make highlighted text a bit darker
- add tiny border-radius to highlighted blocks *blingbling*

muhqu committed:
preserve tabs for inline diff and replace them with spaces after whitespace changes have been highlighted

muhqu commented:
just fixed the tabs-to-spaces issue

atuttle commented:
border-radius, eh? Got an updated screenshot?

muhqu committed:
fix highlighting of block indentions

changed whitespaceAwareTokenize to split at line breaks or spaces/tabs so that tokens of preceding whitespace do not include the last line break

muhqu commented:
@atuttle look here:
![tiny border radius](https://f.cloud.github.com/assets/26716/964961/5916dfbe-0521-11e3-833a-884cae4e6741.png)

muhqu committed:
implement much better tokenization for programming languages.

it splits on:
- every line break
- bunch of whitespace
- bunch of word characters
- every non-word characters (e.g. punctuation)

muhqu commented:
just updated the screenshot in the description to match the latest version of this. From my point of view this is ready for merge, further improvements can be added later with additional pull requests.

Maybe it could make sense to add a User Preference wether to use contrastful or sober diff colors... or just let users define `gitx.diffcolors` using `git config`?

atuttle commented:
I don't feel like the border radius is really adding anything. It's not hurting anything either, but meh.

Hope this is merged in soon!

muhqu commented:
I'm using it now for work now. I doesn't crash but I have some commits where the diff view keeps showing "This is a large commit. Click here or press 'v' to view". Guess I have to investigate this before this is ready to merge...

muhqu committed:
improve tokenizer performance by using single regex  \o/

muhqu commented:
If anyone wants to help test this PR:
a) patch your javascript/css files by hand, e.g. open a terminal window in the Folder where your GitX.app is and type: `open GitX.app/Contents/Resources/html` 
b) install this binary: http://muhqu.de/builds/GitX-dev-DiffWordHighlight@9097c08fc3.zip

atuttle commented:
Installed. I'll see how it works for me today.

atuttle commented:
- Installed via the zip file & overwrote my existing Gitx.app
- Opened by typing `gitx` from the CLI
- Doesn't look like I'm getting any word highlighting:

![](https://www.evernote.com/shard/s240/sh/78eb0d6c-d2e4-4438-96c8-43186a6aec3d/0a7e8c263f0cbe3a3e9ec093ee06a2c5/deep/0/Screenshot%208/22/13%209:27%20AM.png)

muhqu committed:
to be more consistent highlight whole line even though hunk contains only additions or deletions

muhqu commented:
rebased PR: rowanj/master@dda41e4440e549a83d9a9dfcd3a0ded093c8db66 tag: builds/testing/2

atuttle commented:
@muhqu any chance of a new binary download with your latest changes?

rowanj commented:
It does look pretty good; I still need to review it properly but I'm considering merging it and putting out a new testing build soon.

Is the "this is a large commit" a bug in these changes, or are those commits over the existing threshold?

I've been considering either dropping that code, or hugely changing the threshold anyway; I don't think that loading the larger commits is the problem it may have once been.

muhqu commented:
@rowanj the issue with "this is a large commit" did not show up again. I guess it was only happening in a custom build with other branches merged all together... 

You should definitely **not drop** the 'large commit' feature. I think it is fairly useful as it is. Maybe consider adding a NSUserDefault for the threshold.

The 'bonus todo' to customise diff colors will be covered by a separate issue #249

muhqu committed:
fix html evaluation, missing escape

muhqu commented:
@atuttle latest build of this PR: http://muhqu.de/builds/GitX-dev-DiffWordHighlight@05d4872be2.app.zip

rowanj commented:
I don't think the 'large commit' message is actually useful at all; the commit loading should be asynchronous, fast, and cancelled if the user navigates away.  Waiting for user intervention before beginning the load process is a poor interface choice.

The only real point of that screen is to let you know that it might take some amount of time, but the actual time taken is not consistent between machines or even really dependant on the metrics used to show that message.  It should just be a "Loading..." spinner or something.

rowanj commented:
Oh, I do like the border radius - it adds a little visual distinction to changes that start or end at the same column in adjacent lines :-)

muhqu commented:
I have some repositories that include the codebase of other frameworks, which results in fairly large commits when updating these frameworks to new versions. Just selecting such a commit should not result in loads of wasted CPU cycles to generate the whole highlighted diff result of which I only see the first 60 lines at max... Also the memory consumption to generate the DOM could become a bottleneck. You often select commits in the history log not in order to take a look at the diff but to do other things like checkout a branch, or just read a commit message...

I agree that the metric used to show the message is not a great source to make assumptions on how expensive it would be to generate the diff. So it could make sense to change the loading behaviour to show the "This is a large commit" message only if `highlightDiff()` exceeds e.g. 500 ms to load.

atuttle commented:
@muhqu Thanks for the updated binary. I overwrote my existing GitX.app but I'm still not seeing word-diffs...

![](https://www.evernote.com/shard/s240/sh/f760f500-92c4-47ea-83e7-1f4116d30069/03237636b4311c85058357cd6873836c/deep/0/Screenshot%208/27/13%209:44%20AM.png)

muhqu commented:
@atuttle thats odd... you said you start GitX via `gitx` commandline, have you tried reinstalling the command line tool ( "GitX" menu -> "Enable Terminal Usage..." ). Make sure to quit GitX if already running and start specific binary  before doing so.

Unfortunately the "About" dialog doesn't show whether its a release build or a developer build directly from some git revision. @rowanj it would be great if we could get something like `git describe` into this dialog...

One further thing I would like to put into this PR is enhancing file renames. Currently file renames are shown as follows:
- very/long/path/to/old-filename.html -> very/long/path/to/new-filename.html

I would like to change it to something like:
- very/long/path/to/{ old -> new }-filename.html

atuttle commented:
Sigh. I would have sworn I only had 1 GitX install, but I was wrong. I removed all existing binaries, reinstalled the latest from @muhqu and reset the CLI command, and now I see word diffs.

Thanks!

atuttle commented:
Now that I've got it working, @muhqu's build needs more work (sorry!)... Sometimes clicking on a modified file in the stage view shows nothing, sometimes it shows the wrong contents (sometimes it's right...). I'm available to demo this via G+ hangout (adam@fusiongrokker.com) if you want to see what I see.

muhqu committed:
display file renames inline, e.g. very/long/path/to/ { old -> new } -filename.html

muhqu committed:
fix unescaped filenames in commit dialog title

muhqu committed:
fix various inline-diff issues caused by unexpected line breaks

muhqu commented:
@atuttle thanks for testing! I have just pushed a few changes which might already fix you issues (if they relate to line-break issues...). You can find a fresh binary here: http://muhqu.de/builds/GitX-dev-DiffWordHighlight@3717db3.app.zip

If your issue is still present, it would be great if you could send me the corresponding `git diff -M` output.

atuttle commented:
@muhqu thanks for the new build. So far so good in stage view, but I'm still having trouble with viewing history. Sometimes the history shows up as expected, sometimes the filenames are listed but not their contents, and sometimes I get the "loading" graphic, but it doesn't seem to finish within 5-10 sec so I moved on.

atuttle commented:
@muhqu Here's an issue with the stage view... Line numbers are wrong. I can illustrate this by changing the amount of context shown:

With context set to pretty low, this change appears to be on  "line 2":

![](https://www.evernote.com/shard/s240/sh/4bf8bdfd-d298-41a0-a0f1-d7c99cd543a1/ccf714f4f73bfca5843c7edc40dae14d/deep/0/Screenshot%208/29/13%202:22%20PM.png)

But if I bump up the context, it changes lines:

![](https://www.evernote.com/shard/s240/sh/30fb03eb-e1b7-4735-8937-339dc4ef4820/62422f19c26e9c7c71ca7f086b47ab78/deep/0/Screenshot%208/29/13%202:23%20PM.png)

And it gets weirder. If I bump the context _any_ higher than pictured above, the amount of included context goes crazy. This is with the slider just 1 increment higher: (note the visible scrollbar -- there's more above that's out of the viewport!)

![](https://www.evernote.com/shard/s240/sh/fac1cb50-70e8-42c7-a538-b942e28114d6/ab7f63ddd05e2efadcfff23118ca8c81/deep/0/Screen%20Shot%202013-08-29%20at%202.24.53%20PM%20%282%29.png)

atuttle commented:
@muhqu Also, line numbering seems to continue through hunks. For example:

![](https://www.evernote.com/shard/s240/sh/229ef11e-eb72-4ded-9488-88e11bbdec3d/be82a9c56aa1d47724ef6b8115c6ff5e/deep/0/Screenshot%208/29/13%202:29%20PM.png)

"Line 133" here is actually line 117; and because of this bug, the line numbers displayed in GitX can far exceed the number of lines in the file. This particular file is 191 lines, but with context cranked up to maximum GitX shows up to  "line 242"

muhqu commented:
@atuttle thanks a lot for the details. The underlaying diff data comes directly from a call to `git diff`. What `git version` are us using with GitX? ( ...I'm currently using 1.8.3.1 )

atuttle commented:
I'm on 1.8.0. Suppose I could upgrade...

Adam

On Fri, Aug 30, 2013 at 3:22 AM, Mathias Leppich
notifications@github.comwrote:

> @atuttle https://github.com/atuttle thanks a lot for the details. The
> underlaying diff data comes directly from a call to git diff. What git
> version are us using with GitX? ( ...I'm currently using 1.8.3.1 )
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/rowanj/gitx/pull/243#issuecomment-23544558
> .

atuttle commented:
Upgraded to 1.8.3.4 and I'm not having any line-numbering or hunk-size
issues any more.

Adam

On Fri, Aug 30, 2013 at 7:08 AM, Adam Tuttle adam@fusiongrokker.com wrote:

> I'm on 1.8.0. Suppose I could upgrade...
> 
> Adam
> 
> On Fri, Aug 30, 2013 at 3:22 AM, Mathias Leppich <notifications@github.com
> 
> > wrote:
> > 
> > @atuttle https://github.com/atuttle thanks a lot for the details. The
> > underlaying diff data comes directly from a call to git diff. What git
> > version are us using with GitX? ( ...I'm currently using 1.8.3.1 )
> > 
> > —
> > Reply to this email directly or view it on GitHubhttps://github.com/rowanj/gitx/pull/243#issuecomment-23544558
> > .

muhqu commented:
Oh... thanks a lot for clarifying ... I will try to repro with 1.8.0 to identify the issue. It has to be some difference in the diff output thats generated from the cmd-line `git diff`...

atuttle commented:
Can you tell I really want this feature? ;)

Adam

On Fri, Aug 30, 2013 at 9:37 AM, Mathias Leppich
notifications@github.comwrote:

> Oh... thanks a lot for clarifying ... I will try to repro with 1.8.0 to
> identify the issue. It has to be some difference in the diff output thats
> generated from the cmd-line git diff...
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/rowanj/gitx/pull/243#issuecomment-23561371
> .

rowanj commented:
Heh, chalk up another point in the "reasons not to use command-line git output" column :-)

I guess I'll try to parse the Sparkle stats to see how common Git 1.8.0.x is.

muhqu commented:
awesome, that would be quiet interesting!

muhqu commented:
@rowanj yesterday I played around with the 'large commit' handling and found a way to make the diff-loading-process more progressively. All diffs are now handled the same way, a progress bar shows and updates only at 100ms. So most of the time commits are loaded fast enough to not show a progress bar at all.

Here's a screenshot:
![gitx_progressive_diff_highlight](https://f.cloud.github.com/assets/26716/1078538/3a4b564a-152b-11e3-8474-04b609abcc21.png)

Should I add this change to this pull request or do you prefer a separate, rebased pull request?

rowanj commented:
- [ ] Add attribution and disclaimer notice as per MIT license

muhqu commented:
:+1:  awesome! looking fwd to a new beta build! ;-)

atuttle commented:
:clap:

mitio commented:
Wait, does this word diff rendering look right to you:

![wrong word diff rendering](https://cloud.githubusercontent.com/assets/129307/2956410/d0399a80-da8e-11e3-8b23-5dde993d3afc.png)

The screenshot is from Version 0.15.1869 dev (0.15.1869).

muhqu commented:
@mitio huh... that's interesting. I created [this pen](http://codepen.io/anon/pen/BkHrD) which illustrates and helps isolating the issue.

muhqu commented:
so the issue is when text parts just move around...  anyway, this is an issue and I will look into it once I find time for it, thanks for reporting

mitio commented:
No problem. I created #321 since this has already been merged. Let's continue the discussion there.
